High level sequence diagram (#27)
peppelinux/draft-demarco-acme-openid-federation: draft-demarco-acme-openid-federation.md

      Discovery](https://openid.net/specs/openid-federation-1_0.html#section-8)
      to obtain the Trust Chain related to the Requestor.

> The following diagram illustrates a successful interaction between Issuer and
> Requestor to retrieve a certificate. It assumes the Requestor has already
> discovered the Issuer, and has already created an ACME account.

Reviewer: ```suggestion
The following diagram illustrates a successful interaction between Issuer and
Requestor to retrieve an X.509 Certificate. In the diagram there is flow about a Requestor that has already
discovered the Issuer
```

it is not clear to me the presence of this requirement about the acme account creation.
Using an automatic issuance the acme account would not be required, or differently, it would be created during the first interaction between a requestor and a issuer, on the fly.

We need to further align on this element.

Author: I will note the section of ACME that relates to the account.

> TODO: convert mermaid diagram to ASCII before submitting as ID (or does data
> tracker support mermaid now?).

Reviewer: ```suggestion
```

I will move this seq diagram in a file within the folder called `diagrams` and convert it to ascii arts

Reviewer 2: RFC XML supports a `src` attribute on `<figure>` elements to provide a separate graphics file, but we'll still need an ASCII art text representation to fall back to. `kramdown-rfc2629` supports [running a number of tools](https://github.com/cabo/kramdown-rfc/blob/18d5a95dee9ea9d180a37d8c8a4c5ce1b806339f/lib/kramdown-rfc2629.rb#L581) to turn inline diagram DSLs into SVGs, including `aasvg`, `plantuml`, etc., but not Mermaid, it looks like.

```mermaid
sequenceDiagram
  participant R as Requestor
  participant A as Issuer

  R ->> A: POST /acme/new-order
  A -->> R: Authorization at /acme/authz/[authz-id], Finalize at /acme/order/[order-id]/finalize
  R ->> A: POST /acme/authz/[authz-id]
  A -->> R: openid-federation-01 Challenge at /acme/chall/[chall-id]
  R ->> R: Sign challenge token with private key
  R ->> A: POST /acme/chall/[chall-id] with signed token<br>and entity ID set to Requestor's ID

>   A ->> R: GET /.well-known/openid-federation

Reviewer 2: Would it make sense to split the Requestor participant in this diagram in two, into the Requestor-as-ACME client and the server that hosts the Requestor's OpenID Federation entity?

  R -->> A: Requestor's Entity Configuration
  A ->> A: Check entity configuration sub matches<br>entity identifier in the order
  A ->> A: Check challenge sig is signed with key in<br>entity configuration
  opt If requestor did not provide Trust Chain
    create participant F as Federation Trust Anchors<br>and Intermediates

>     A <<->> F: Determine trust chain between<br>Issuer and Requestor<br>(OpenID Federation Discovery)

Reviewer 2: Since trust chains are asymmetric, I recommend changing this to "Determine trust chain from Issuer's trust anchors to Requestor".

    Discovery](https://openid.net/specs/openid-federation-1_0.html#section-8)
    to obtain the Trust Chain related to the Requestor.

The following diagram illustrates a successful interaction between Issuer and
Requestor to retrieve an X.509 Certificate. The diagram assumes the Requestor
has already discovered the Issuer, and the Requestor has already created an
ACME account with the Issuer.

```mermaid
sequenceDiagram
  participant RF as Requestor's OpenID Federation<br> Web Server
  participant RC as Requestor's ACME Client
  participant A as Issuer

  RC ->> A: POST /acme/new-order
  A -->> RC: Authorization at /acme/authz/[authz-id], Finalize at /acme/order/[order-id]/finalize
  RC ->> A: POST /acme/authz/[authz-id]
  A -->> RC: openid-federation-01 Challenge at /acme/chall/[chall-id]
  RC ->> RC: Sign challenge token with private key
  RC ->> A: POST /acme/chall/[chall-id] with signed token<br>and entity ID set to Requestor's ID
  A ->> RF: GET /.well-known/openid-federation
  RF -->> A: Requestor's Entity Configuration

>   A ->> A: Check entity configuration sub matches<br>entity identifier in the order

Reviewer: ```suggestion
  A ->> A: Check Entity Configuration sub matches<br>Entity identifier in the order
```

>   A ->> A: Check challenge sig is signed with key in<br>entity configuration

Reviewer: ```suggestion
  A ->> A: Check challenge sig is signed with key in<br>Entity Configuration
```

  opt If requestor did not provide Trust Chain
    create participant F as Federation Trust Anchors<br>and Intermediates

>     A <<->> F: Determine trust chain from<br>Issuer's trust anchors to Requestor<br>(OpenID Federation Discovery)

Reviewer: ```suggestion
    A <<->> F: Determine Trust Chain from<br>Issuer's Trust Anchors to Requestor<br>(OpenID Federation Discovery)
```

  end
  A ->> A: Evaluate trust chain
  A -->> RC: Respond to POST with validation success
  RC ->> A: POST /acme/orders/[order-id]/finalize with CSR
  A ->> A: Check CSR validity according to protocol and CA policy
  A -->> RC: Order object with certificate at /acme/cert/[cert-id]
  RC ->> A: POST /acme/cert/[cert-id]

>   A -->> RC: Newly issued certificate 🎉

Reviewer: ```suggestion
  A -->> RC: Newly issued X.509 Certificate 🎉
```
